sklearn/base.py

```diff
@@ -342,6 +344,71 @@ def _get_tags(self):
                 collected_tags.update(more_tags)
         return collected_tags
 
+    def _check_n_features(self, X, reset):
+        """Set the `n_features_in_` attribute, or check against it.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples, n_features)
+            The input samples.
+        reset : bool
+            If True, the `n_features_in_` attribute is set to `X.shape[1]`.
+            Else, the attribute must already exist and the function checks
+            that it is equal to `X.shape[1]`.
+        """
+        n_features = X.shape[1]
+
+        if reset:
+            self.n_features_in_ = n_features
+        else:
+            if not hasattr(self, 'n_features_in_'):
+                raise RuntimeError(
+                    "The reset parameter is False but there is no "
+                    "n_features_in_ attribute. Is this estimator fitted?"
+                )
+            if n_features != self.n_features_in_:
+                raise ValueError(
+                    'X has {} features, but this {} is expecting {} features '
+                    'as input.'.format(n_features, self.__class__.__name__,
+                                       self.n_features_in_)
+                )
+
+    def _validate_data(self, X, y=None, reset=True, **check_params):
+        """Validate input data and set or check the `n_features_in_` attribute.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix, dataframe} of shape \
+                (n_samples, n_features)
+            The input samples.
+        y : array-like of shape (n_samples,), default=None
+            The targets. If None, `check_array` is called on `X` and
+            `check_X_y` is called otherwise.
+        reset : bool, default=True
+            Whether to reset the `n_features_in_` attribute. See
+            :func:`_check_n_features`.
+        **check_params : kwargs
+            Parameters passed to :func:`sklearn.utils.check_array` or
+            :func:`sklearn.utils.check_X_y`.
+
+        Returns
+        -------
+        out : {ndarray, sparse matrix} or tuple of these
+            The validated input. A tuple is returned if `y` is not None.
+        """
+
+        if y is None:
+            X = check_array(X, **check_params)
+            out = X
+        else:
+            X, y = check_X_y(X, y, **check_params)
+            out = X, y
```

Comment:
At first glance, I'm not sure of the benefits?

It would just be a less obvious way to call `check_array` so that goes against the Zen "There should be one-- and preferably only one --obvious way to do it."

Reply:
The main benefit would be to have an explicit and standard `check_params` for estimators to specify when they accept `None` as valid value for `y` (e.g. for most unsupervised transformers) and when `y=None` is not valid and should raise a `ValueError` (e.g. for all classifiers and regressors).

Reply:
Wouldn't it be better directly in `_validate_data` then? As proposed in https://github.com/scikit-learn/scikit-learn/pull/16112#discussion_r376441226

Reply:
I'm not sure I follow @ogrisel.  Currently, `check_X_y` raises an error when `y` is None. Are you suggesting to add a parameter to make this optional (both in `check_X_y` and in `_validate_data`)? In `_validate_data` this would be a bit strange because the call would be `X = _validate_data(X, y_ignored=True)` and `y` isn't passed anyway.

Reply:
I don't think we need to change check_X_y in this PR. There is enough to review here